iot_config/src/client/mod.rs

```diff
@@ -0,0 +1,136 @@
+};
+use std::{sync::Arc, time::Duration};
+
+mod settings;
+pub use settings::Settings;
+
+#[derive(thiserror::Error, Debug)]
+pub enum ClientError {
+    #[error("error signing request: {0}")]
+    Signing(#[from] helium_crypto::Error),
+    #[error("grpc error response: {0}")]
+    Rpc(#[from] tonic::Status),
+    #[error("error verifying response signature: {0}")]
+    Verification(#[from] file_store::Error),
+    #[error("error resolving region params: {0}")]
+    UndefinedRegionParams(String),
+}
+
+#[derive(Clone, Debug)]
+pub struct Client {
+    pub gateway_client: iot_config::gateway_client::GatewayClient<Channel>,
+    pub admin_client: iot_config::admin_client::AdminClient<Channel>,
+    signing_key: Arc<Keypair>,
+    config_pubkey: PublicKey,
+    batch_size: u32,
+}
+
+impl Client {
+    pub fn from_settings(settings: &Settings) -> Result<Self, Box<helium_crypto::Error>> {
+        let channel = Endpoint::from(settings.url.clone())
+            .connect_timeout(Duration::from_secs(settings.connect_timeout))
+            .timeout(Duration::from_secs(settings.rpc_timeout))
+            .connect_lazy();
+        Ok(Self {
+            gateway_client: iot_config::gateway_client::GatewayClient::new(channel.clone()),
+            admin_client: iot_config::admin_client::AdminClient::new(channel),
+            signing_key: settings.signing_keypair()?,
+            config_pubkey: settings.config_pubkey()?,
+            batch_size: settings.batch_size,
+        })
+    }
+
+    pub async fn resolve_region_params(
+        &mut self,
+        region: Region,
+    ) -> Result<RegionParamsInfo, ClientError> {
+        let mut request = iot_config::RegionParamsReqV1 {
+            region: region.into(),
+            signer: self.signing_key.public_key().into(),
+            signature: vec![],
+        };
+        request.signature = self.signing_key.sign(&request.encode_to_vec())?;
+        let response = self.admin_client.region_params(request).await?.into_inner();
+        response.verify(&self.config_pubkey)?;
+        Ok(RegionParamsInfo {
+            region: response.region(),
+            region_params: response
+                .params
+                .ok_or_else(|| ClientError::UndefinedRegionParams(format!("{region}")))?
+                .region_params,
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl gateway_info::GatewayInfoResolver for Client {
+    type Error = ClientError;
+
+    async fn resolve_gateway_info(
+        &mut self,
+        address: &PublicKeyBinary,
+    ) -> Result<Option<gateway_info::GatewayInfo>, Self::Error> {
+        let mut request = iot_config::GatewayInfoReqV1 {
+            address: address.clone().into(),
+            signer: self.signing_key.public_key().into(),
+            signature: vec![],
+        };
+        request.signature = self.signing_key.sign(&request.encode_to_vec())?;
+        tracing::debug!(pubkey = address.to_string(), "fetching gateway info");
```

Comment:
nit: not sure what value this log statement provide

Reply:
this is probably a better place for a metric on grpc requests then a log statement, good point